File: src/55_02_BalancedBinaryTree/55_02_BalancedBinaryTree.cpp
// 面试题55（二）：平衡二叉树
// 题目：输入一棵二叉树的根结点，判断该树是不是平衡二叉树。如果某二叉树中
// 任意结点的左右子树的深度相差不超过1，那么它就是一棵平衡二叉树。

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 * };
 */

#include <algorithm>
#include <cmath>
#include <iostream>
#include <string>
#include <vector>

#include "../Utilities/BinaryTree.h"

using namespace std;

using TreeNode = BinaryTreeNode;

class Solution {
   public:
    bool isBalanced( TreeNode* root ) {
        return recur( root ) != -1;  // 不平衡标志为-1
    }

   private:
    // 后序遍历+剪枝。从底至顶判断左右子树是否平衡。
    // 返回当前子树的深度，或者不平衡标志-1
    int recur( TreeNode* root ) {
        // 终止条件：到达叶节点，返回深度为0
        if ( !root ) return 0;
        // 判断左子树
        int left = recur( root->left );
        if ( left == -1 ) return -1;  // 如果左子树不是平衡树则提前结束（剪枝）
        // 判断右子树
        int right = recur( root->right );
        if ( right == -1 ) return -1;
        // 左右子树均平衡，判断当前是否平衡。
        // 若是则返回当前深度 (子树深度+1)，否则返回-1
        return std::abs( left - right ) < 2 ? std::max( left, right ) + 1 : -1;
    }
};

class Test {
   public:
    void checkResult( const string testName, TreeNode* root, bool expected ) {
        cout << testName;
        bool result = solution.isBalanced( root );
        if ( result == expected )
            cout << " passed.\n";
        else
            cout << " failed.\n";
    }

    // 完全二叉树
    //             1
    //         /      \
    //        2        3
    //       /\       / \
    //      4  5     6   7
    void Test1() {
        ValVec vals{ 1, 2, 3, 4, 5, 6, 7 };
        NodeVec nodes = createNodesVec( vals );

        connectNodesIndex( nodes, 0, 1, 2 );
        connectNodesIndex( nodes, 1, 3, 4 );
        connectNodesIndex( nodes, 2, 5, 6 );

        checkResult( "Test1", nodes[ 0 ], true );
        destroyTree( nodes[ 0 ] );
    }

    // 不是完全二叉树，但是平衡二叉树
    //             1
    //         /      \
    //        2        3
    //       /\         \
    //      4  5         6
    //        /
    //       7
    void Test2() {
        ValVec vals{ 1, 2, 3, 4, 5, 6, 7 };
        NodeVec nodes = createNodesVec( vals );

        connectNodesIndex( nodes, 0, 1, 2 );
        connectNodesIndex( nodes, 1, 3, 4 );
        connectNodesIndex( nodes, 2, -1, 5 );
        connectNodesIndex( nodes, 4, 6, -1 );

        checkResult( "Test2", nodes[ 0 ], true );
        destroyTree( nodes[ 0 ] );
    }

    // 不是平衡二叉树
    //             1
    //         /      \
    //        2        3
    //       /\         
    //      4  5
    //        /
    //       6
    void Test3() {
        ValVec vals{ 1, 2, 3, 4, 5, 6 };
        NodeVec nodes = createNodesVec( vals );

        connectNodesIndex( nodes, 0, 1, 2 );
        connectNodesIndex( nodes, 1, 3, 4 );
        connectNodesIndex( nodes, 4, 5, -1 );

        checkResult( "Test3", nodes[ 0 ], false );
        destroyTree( nodes[ 0 ] );
    }

    //               1
    //              /
    //             2
    //            /
    //           3
    //          /
    //         4
    //        /
    //       5
    void Test4() {
        ValVec vals{ 1, 2, 3, 4, 5 };
        NodeVec nodes = createNodesVec( vals );

        connectNodesIndex( nodes, 0, 1, -1 );
        connectNodesIndex( nodes, 1, 2, -1 );
        connectNodesIndex( nodes, 2, 3, -1 );
        connectNodesIndex( nodes, 3, 4, -1 );

        checkResult( "Test4", nodes[ 0 ], false );
        destroyTree( nodes[ 0 ] );
    }

    // 1
    //  \
    //   2
    //    \
    //     3
    //      \
    //       4
    //        \
    //         5
    void Test5() {
        ValVec vals{ 1, 2, 3, 4, 5 };
        NodeVec nodes = createNodesVec( vals );

        connectNodesIndex( nodes, 0, -1, 1 );
        connectNodesIndex( nodes, 1, -1, 2 );
        connectNodesIndex( nodes, 2, -1, 3 );
        connectNodesIndex( nodes, 3, -1, 4 );

        checkResult( "Test5", nodes[ 0 ], false );
        destroyTree( nodes[ 0 ] );
    }

    // 树中只有1个结点
    void Test6() {
        ValVec vals{ 1 };
        NodeVec nodes = createNodesVec( vals );

        checkResult( "Test6", nodes[ 0 ], true );
        destroyTree( nodes[ 0 ] );
    }

    // 树中没有结点
    void Test7() { checkResult( "Test7", nullptr, true ); }

    void runTest() {
        Test1();
        Test2();
        Test3();
        Test4();
        Test5();
        Test6();
        Test7();
    }

   private:
    Solution solution;
};

int main() {
    Test test;
    test.runTest();
    return 0;
}

File: src/55_02_BalancedBinaryTree/55_02_BalancedBinaryTree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

import os
import sys

sys.path.append(os.path.split(sys.path[0])[0])


from Utilities import binary_tree
from Utilities.binary_tree import TreeNode


class Solution:
    def isBalanced(self, root: TreeNode) -> bool:
        return self.recur(root) != -1

    # 后序遍历+剪枝。从底至顶判断左右子树是否平衡。
    # 返回当前子树的深度，或者不平衡标志-1
    def recur(self, root: TreeNode) -> int:
        if not root:
            return 0
        # 判断左子树
        left = self.recur(root.left)
        if left == -1:  # 如果左子树不是平衡树则提前结束（剪枝）
            return -1
        # 判断右子树
        right = self.recur(root.right)
        if right == -1:
            return -1
        # 左右子树均平衡，判断当前是否平衡。
        # 若是则返回当前深度 (子树深度+1)，否则返回-1
        return max(left, right) + 1 if abs(left - right) < 2 else -1


class Test:
    def __init__(self) -> None:
        self.solution = Solution()

    def checkResult(self, testName: str, root: TreeNode, expected: bool) -> None:
        text = testName
        result = self.solution.isBalanced(root)
        if result == expected:
            text += " passed."
        else:
            text += " failed."
        print(text)

    # 完全二叉树
    #             1
    #         /      \
    #        2        3
    #       /\       / \
    #      4  5     6   7
    def Test1(self):
        vals = [1, 2, 3, 4, 5, 6, 7]
        nodes = binary_tree.createTreeNodesVec(vals)

        idxs = [
            (0, 1, 2),
            (1, 3, 4),
            (2, 5, 6),
        ]
        binary_tree.connectTreeNodesIndexVec(nodes, idxs)

        self.checkResult("Test1", nodes[0], True)

    # 不是完全二叉树，但是平衡二叉树
    #             1
    #         /      \
    #        2        3
    #       /\         \
    #      4  5         6
    #        /
    #       7
    def Test2(self):
        vals = [1, 2, 3, 4, 5, 6, 7]
        nodes = binary_tree.createTreeNodesVec(vals)

        idxs = [
            (0, 1, 2),
            (1, 3, 4),
            (2, -1, 5),
            (4, 6, -1),
        ]
        binary_tree.connectTreeNodesIndexVec(nodes, idxs)

        self.checkResult("Test2", nodes[0], True)

    # 不是平衡二叉树
    #             1
    #         /      \
    #        2        3
    #       /\
    #      4  5
    #        /
    #       6
    def Test3(self):
        vals = [1, 2, 3, 4, 5, 6]
        nodes = binary_tree.createTreeNodesVec(vals)

        idxs = [
            (0, 1, 2),
            (1, 3, 4),
            (4, 5, -1),
        ]
        binary_tree.connectTreeNodesIndexVec(nodes, idxs)

        self.checkResult("Test3", nodes[0], False)

    #               1
    #              /
    #             2
    #            /
    #           3
    #          /
    #         4
    #        /
    #       5
    def Test4(self):
        vals = [1, 2, 3, 4, 5]
        nodes = binary_tree.createTreeNodesVec(vals)

        idxs = [
            (0, 1, -1),
            (1, 2, -1),
            (2, 3, -1),
            (3, 4, -1),
        ]
        binary_tree.connectTreeNodesIndexVec(nodes, idxs)

        self.checkResult("Test4", nodes[0], False)

    # 1
    #  \
    #   2
    #    \
    #     3
    #      \
    #       4
    #        \
    #         5
    def Test5(self):
        vals = [1, 2, 3, 4, 5]
        nodes = binary_tree.createTreeNodesVec(vals)

        idxs = [
            (0, -1, 1),
            (1, -1, 2),
            (2, -1, 3),
            (3, -1, 4),
        ]
        binary_tree.connectTreeNodesIndexVec(nodes, idxs)

        self.checkResult("Test5", nodes[0], False)

    # 树中只有1个结点
    def Test6(self):
        vals = [1]
        nodes = binary_tree.createTreeNodesVec(vals)

        self.checkResult("Test6", nodes[0], True)

    # 树中没有结点
    def Test7(self):
        self.checkResult("Test7", None, True)

    def runTest(self):
        self.Test1()
        self.Test2()
        self.Test3()
        self.Test4()
        self.Test5()
        self.Test6()
        self.Test7()


if __name__ == "__main__":
    test = Test()
    test.runTest()
